### PR TITLE
Move implementations from SDK Sender and Receiver

### PR DESCRIPTION
**Pull request by conradgrobler**
Move the context-dependent implementations from the SDK's `Sender` and `Receiver` structs into extension traits.

This is a first step towards implementing #1224, #1249, #1324, and generally making it possible to serialise handles in protos in the runtime pseudo nodes as well. This requires harmonising the `Sender` and `Receiver` structs between the SDK and runtime. Some of the functions are incompatible between the runtime and SDK, so have to be moved out of the implementation of the struct itself. The next step is to do something similar for the runtime versions of these structs.

- `SenderExt`
  - `send`
  - `close`
- `ReceiverExt`
  - `receive`
  - `wait`
  - `try_receive`
  - `close`

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: examples/abitest/module_0/rust/src/lib.rs [hunk 1]
```diff
@@ -1869,7 +1868,7 @@ impl FrontendNode {
             .map_err(from_proto)?;
         let mut count = 0;
         loop {
-            match receiver.receive() {
+            match oak::io::receive(&receiver) {
```
Comment (tiziano88) [inline examples/abitest/module_0/rust/src/lib.rs:1871]: I think for the SDK crate we can actually recover most of the existing semantics by declaring an extension that exposes the `.receive` and other methods directly on the sender / receiver types (the usual pattern is to have a `ReceiverExt` and `SenderExt` trait, but the naming does not really matter).

Comment (conradgrobler) [reply examples/abitest/module_0/rust/src/lib.rs]: I changed it to use extension traits as you suggested.

File: examples/injection/module/rust/src/lib.rs [hunk 2]
```diff
@@ -66,7 +66,7 @@ mod proto {
 
 use oak::{
     grpc,
-    io::{Receiver, Sender},
+    io::{Receiver, ReceiverExt, Sender, SenderExt},
```
Comment (tiziano88) [inline examples/injection/module/rust/src/lib.rs:69]: I imagine eventually `oak::io::Receiver` and `oak::io::Sender` would be replaced by simpler objects (possibly with same name) defined in `oak_abi` , and also generated by our custom prost logic, is that correct?

Comment (conradgrobler) [reply examples/injection/module/rust/src/lib.rs:69]: Yes, although it would probably move to the new `oak_io` crate as part of #1324.

**Pull request merged**